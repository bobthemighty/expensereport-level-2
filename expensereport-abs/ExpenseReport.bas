DINNER = 0: BREAKFAST = 1: CARRENTAL = 2
DIM type(5), amount(5)
type(0) = DINNER: amount(0) = 5000
type(1) = DINNER: amount(1) = 5001
type(2) = BREAKFAST: amount(2) = 1000
type(3) = BREAKFAST: amount(3) = 1001
type(4) = CARRENTAL: amount(4) = 4
htmlMode = 1

GOSUB PrintExpenses
END

PrintExpenses:
    meals = 0
    total = 0
    IF htmlMode THEN
        PRINT "<!DOCTYPE html>"
        PRINT "<html lang=" CHR$(34) "en" CHR$(34) ">"
        PRINT "<head>"
        PRINT "<title>Expense Report " DATE$ "</title>"
        PRINT "</head>"
        PRINT "<body>"
        PRINT "<h1>Expenses " DATE$ "</h1>"
    ELSE
        PRINT "Expensereport: " DATE$
    END IF

    IF htmlMode THEN
        PRINT "<table>"
        PRINT "<thead>"
        PRINT "<tr><th scope=" CHR$(34) "col" CHR$(34) ">Type</th><th scope=" CHR$(34) "col" CHR$(34) ">Amount</th><th scope=" CHR$(34) "col" CHR$(34) ">Over Limit</th></tr>"
        PRINT "</thead>"
        PRINT "<tbody>"
    END IF

    FOR i=0 TO 4
        IF type(i) = DINNER OR type(i) = BREAKFAST THEN meals = meals + amount(i)
        expenseName$ = ""
        IF type(i) = DINNER THEN expenseName$ = "Dinner"
        IF type(i) = BREAKFAST THEN expenseName$ = "Breakfast"
        IF type(i) = CARRENTAL THEN expenseName$ = "Car Rental"
        IF type(i) = DINNER AND amount(i) > 5000 OR type(i) = BREAKFAST AND amount(i) > 1000 THEN mealOverExpensesMarker$ = "X" ELSE mealOverExpensesMarker$ = " "
        IF htmlMode THEN
            PRINT "<tr><td>" expenseName$ "</td><td>" STR$(amount(i)) "</td><td>" mealOverExpensesMarker$ "</td></tr>"
        ELSE
            PRINT expenseName$ CHR$(9) STR$(amount(i)) CHR$(9) mealOverExpensesMarker$
        END IF
        total = total + amount(i)

    NEXT i

    IF htmlMode THEN
        PRINT "</tbody>"
        PRINT "</table>"
    END IF

    IF htmlMode THEN
        PRINT "<p>Meal expenses:" STR$(meals) "</p>"
        PRINT "<p>Total expenses:" STR$(total) "</p>"
    ELSE
        PRINT "Meal Expenses:" STR$(meals)
        PRINT "Total Expenses:" STR$(total)
    END IF

    IF htmlMode THEN
        PRINT "</body>"
        PRINT "</html>"
    END IF
    RETURN
