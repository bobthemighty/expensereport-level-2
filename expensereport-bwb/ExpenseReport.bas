10 DINNER = 0: BREAKFAST = 1: CARRENTAL = 2
20 DIM TYPE(4): DIM AMOUNT(4)
30 TYPE(0) = DINNER: AMOUNT(0) = 5000
40 TYPE(1) = DINNER: AMOUNT(1) = 5001
50 TYPE(2) = BREAKFAST: AMOUNT(2) = 1000
60 TYPE(3) = BREAKFAST: AMOUNT(3) = 1001
70 TYPE(4) = CARRENTAL: AMOUNT(4) = 4
80 htmlMode = 1
90 GOSUB 1000
100 QUIT
1000 REM PRINTREPORT
1010 MEALS = 0: SUM = 0
1011 IF htmlMode = 0 THEN GOTO 1020
1011 PRINT "<!DOCTYPE html>"
1012 PRINT "<html lang="; CHR$(34); "en"; CHR$(34); ">"
1013 PRINT "<head>"
1014 PRINT "<title>Expense Report</title>"
1015 PRINT "</head>"
1016 PRINT "<body>"
1017 PRINT "<h1>Expense Report</h1>"
1018 GOTO 1021
1020 PRINT "Expense Report:"
1021 IF htmlMode = 0 THEN GOTO 1030
1022 PRINT "<table>"
1023 PRINT "<thead>"
1024 PRINT "<tr><th scope="; CHR$(34); "col"; CHR$(34); ">Type</th><th scope="; CHR$(34); "col"; CHR$(34); ">Amount</th><th scope="; CHR$(34); "col"; CHR$(34); ">Over Limit</th></tr>"
1025 PRINT "</thead>"
1026 PRINT "<tbody>"
1030 FOR I = 0 TO 4
1040 IF TYPE(I) = 0 OR TYPE(I) = 1 THEN MEALS = MEALS + AMOUNT(I)
1050 NAME$ = ""
1060 IF TYPE(I) = 0 THEN NAME$ = "Dinner"
1070 IF TYPE(I) = 1 THEN NAME$ = "Breakfast"
1080 IF TYPE(I) = 2 THEN NAME$ = "Car Rental"
1090 IF TYPE(I) = 0 AND AMOUNT(I) > 5000 THEN GOTO 1130
1100 IF TYPE(I) = 1 AND AMOUNT(I) > 1000 THEN GOTO 1130
1110 MARKER$ = " "
1120 GOTO 1140
1130 MARKER$ = "X"
1140 IF htmlMode = 0 THEN GOTO 1145
1141 PRINT "<tr><td>"; NAME$; "</td><td>"; AMOUNT(I); "</td><td>"; MARKER$; "</td></tr>"
1142 GOTO 1150
1145 PRINT NAME$, AMOUNT(I), MARKER$
1150 SUM = SUM + AMOUNT(I)
1160 NEXT I
1161 IF htmlMode = 0 THEN GOTO 1165
1162 PRINT "</tbody>"
1163 PRINT "</table>"
1165 IF htmlMode = 0 THEN GOTO 1170
1166 PRINT "<p>Meal Expenses: "; MEALS; "</p>"
1167 PRINT "<p>Total Expenses: "; SUM; "</p>"
1168 GOTO 1190
1170 PRINT "Meal Expenses: "; MEALS
1180 PRINT "Total Expenses: "; SUM
1190 IF htmlMode = 0 THEN GOTO 1200
1191 PRINT "</body>"
1192 PRINT "</html>"
1200 RETURN
