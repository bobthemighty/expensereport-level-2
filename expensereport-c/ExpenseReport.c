#include <stdbool.h>
#include <stdio.h>
#include <time.h>

enum Type {
    DINNER,
    BREAKFAST,
    CAR_RENTAL
};

struct Expense {
    enum Type type;
    int amount;
};

void printExpenses(struct Expense expenses[], size_t numExpenses, bool htmlMode) {
    int total = 0;
    int mealExpenses = 0;

    time_t now;
    if (time(&now) == -1)
        return;

    if (htmlMode) {
        printf("<!DOCTYPE html>\n");
        printf("<html lang=\"en\">\n");
        printf("<head>\n");
        printf("<title>Expense Report %s</title>\n", ctime(&now));
        printf("</head>\n");
        printf("<body>\n");
        printf("<h1>Expense Report %s</h1>\n", ctime(&now));
    } else {
        printf("Expenses %s", ctime(&now));
    }

    if (htmlMode) {
        printf("<table>\n");
        printf("<tbody>\n");
        printf("<tr><th span=\"col\">Type</th><th span=\"col\">Amount</th><th span=\"col\">Over Limit</th></tr>\n");
    }
    for (size_t i = 0; i < numExpenses; i++) {
        struct Expense expense = expenses[i];

        if (expense.type == DINNER || expense.type == BREAKFAST) {
            mealExpenses += expense.amount;
        }

        char *expenseName;
        switch (expense.type) {
        case DINNER:
            expenseName = "Dinner";
            break;
        case BREAKFAST:
            expenseName = "Breakfast";
            break;
        case CAR_RENTAL:
            expenseName = "Car Rental";
            break;
        }

        char *mealOverExpensesMarker = ((expense.type == DINNER && expense.amount > 5000) || (expense.type == BREAKFAST && expense.amount > 1000)) ? "X" : " ";

        if (htmlMode) {
            printf("<tr><td>%s</td><td>%d</td><td>%s</td></tr>\n", expenseName, expense.amount, mealOverExpensesMarker);
        } else {
            printf("%s\t%d\t%s\n", expenseName, expense.amount, mealOverExpensesMarker);
        }
        total += expense.amount;
    }
    if (htmlMode) {
        printf("</tbody>\n");
        printf("</table>\n");
    }

    if (htmlMode) {
        printf("<p>Meal expenses: %d</p>\n", mealExpenses);
        printf("<p>Total expenses: %d</p>\n", total);
    } else {
        printf("Meal expenses: %d\n", mealExpenses);
        printf("Total expenses: %d\n", total);
    }

    if (htmlMode) {
        printf("</body>\n");
        printf("</html>\n");
    }
}
