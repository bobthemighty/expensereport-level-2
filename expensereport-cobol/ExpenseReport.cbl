       IDENTIFICATION DIVISION.
       PROGRAM-ID. EXPENSE-REPORT.

       DATA DIVISION.
            WORKING-STORAGE SECTION.
            01 TOTAL PIC 9(10) VALUE 0.
            01 MEALS PIC 9(10) VALUE 0.
            01 EXPENSENAME PIC A(11).
            01 MOEMARKER PIC A(1).
            01 WS-TABLE.
                05 WS-EXPENSES OCCURS 5 TIMES INDEXED BY I.
                    10 WS-TYPE PIC 9(1).
                    10 WS-AMOUNT PIC 9(10).
            01 FORMATTED-INT PIC Z(04)9.
            01 HTML-MODE PIC 9(1) VALUE 1.

       PROCEDURE DIVISION.
           MOVE 1 TO WS-TYPE(1)
           MOVE 1 TO WS-TYPE(2)
           MOVE 2 TO WS-TYPE(3)
           MOVE 2 TO WS-TYPE(4)
           MOVE 3 TO WS-TYPE(5)
           MOVE 5000 TO WS-AMOUNT(1)
           MOVE 5001 TO WS-AMOUNT(2)
           MOVE 1000 TO WS-AMOUNT(3)
           MOVE 1001 TO WS-AMOUNT(4)
           MOVE 4 TO WS-AMOUNT(5)
           PERFORM PRINTREPORT
           STOP RUN.

       PRINTREPORT.
           IF HTML-MODE = 1
               DISPLAY '<!DOCTYPE html>'
               DISPLAY '<html lang="en">'
               DISPLAY '<head>'
               DISPLAY '<title>Expense Report</title>'
               DISPLAY '</head>'
               DISPLAY '<body>'
               DISPLAY '<h1>Expense Report</h1>'
           ELSE
               DISPLAY 'Expenses: '
           END-IF.

           IF HTML-MODE = 1
               DISPLAY '<table>'
               DISPLAY '<thead>'
               DISPLAY '<tr>'
               DISPLAY '<th scope="col">Type</th>'
               DISPLAY '<th scope="col">Amount</th>'
               DISPLAY '<th scope="col">Over Limit</th>'
               DISPLAY '</tr>'
               DISPLAY '</thead>'
               DISPLAY '<tbody>'
           END-IF
           MOVE 1 TO I
           PERFORM SHOWEXPENSEDETAIL
           IF HTML-MODE = 1
               DISPLAY '</tbody>'
               DISPLAY '</table>'
           END-IF.
           MOVE MEALS TO FORMATTED-INT
           IF HTML-MODE = 1
               DISPLAY "<p>Meals: "FORMATTED-INT"</p>"
           ELSE
               DISPLAY "Meals: "FORMATTED-INT
           END-IF.
           MOVE TOTAL TO FORMATTED-INT
           IF HTML-MODE = 1
               DISPLAY "<p>Total: "FORMATTED-INT"</p>"
           ELSE
               DISPLAY "Total: "FORMATTED-INT
           END-IF.
           IF HTML-MODE = 1
               DISPLAY '</body>'
               DISPLAY '</html>'
           END-IF.

       SHOWEXPENSEDETAIL.
           IF WS-TYPE(I) = 1 OR 2
               ADD WS-AMOUNT(I) TO MEALS
           END-IF
           EVALUATE WS-TYPE(I)
               WHEN 1 MOVE 'Dinner'      TO EXPENSENAME
               WHEN 2 MOVE 'Breakfast'   TO EXPENSENAME
               WHEN 3 MOVE 'Car Rental'  TO EXPENSENAME
           END-EVALUATE.
           IF WS-TYPE(I) = 1 AND WS-AMOUNT(I) > 5000
           OR WS-TYPE(I) = 2 AND WS-AMOUNT(I) > 1000
               MOVE 'X' TO MOEMARKER
           ELSE
               MOVE ' ' TO MOEMARKER
           END-IF.
           MOVE WS-AMOUNT(I) TO FORMATTED-INT
           IF HTML-MODE = 1
               DISPLAY "<tr>"
               DISPLAY "<td>"EXPENSENAME"</td>"
               DISPLAY "<td>"FORMATTED-INT"</td>"
               DISPLAY "<td>"MOEMARKER"</td>"
               DISPLAY "</tr>"
           ELSE
               DISPLAY EXPENSENAME FORMATTED-INT ' ' MOEMARKER
           END-IF
           ADD WS-AMOUNT(I) TO TOTAL
           IF I < 5
               ADD 1 TO I
               PERFORM SHOWEXPENSEDETAIL
           END-IF.
