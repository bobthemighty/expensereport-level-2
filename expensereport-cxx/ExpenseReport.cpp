#include <chrono>
#include <iostream>
#include <iterator>
#include <list>

using namespace std;

enum Type
{
    BREAKFAST, DINNER, CAR_RENTAL
};

class Expense
{
    public:
    Type type;
    int amount;
};

void printReport(list<Expense> expenses, bool htmlMode)
{
    int total = 0;
    int mealExpenses = 0;

    auto now = chrono::system_clock::to_time_t(chrono::system_clock::now());

    if (htmlMode) {
        cout << "<!DOCTYPE html>" << '\n';
        cout << "<html lang=\"en\">" << '\n';
        cout << "<head>" << '\n';
        cout << "<title>Expense Report " << ctime(&now) << "</title>\n";
        cout << "</head>" << '\n';
        cout << "<body>" << '\n';
    } else {
        cout << "Expenses " << ctime(&now);
    }

    if (htmlMode) {
        cout << "<table>" << '\n';
        cout << "<thead>" << '\n';
        cout << "<tr><th scope=\"col\">Type</th><th scope=\"col\">Amount</th><th scope=\"col\">Over Limit</th></tr>" << '\n';
        cout << "</thead>" << '\n';
        cout << "<tbody>" << '\n';
    }
    for (list<Expense>::iterator expense = expenses.begin(); expense != expenses.end(); ++expense) {
        if (expense->type == BREAKFAST || expense->type == DINNER) {
            mealExpenses += expense->amount;
        }

        string expenseName = "";
        switch (expense->type) {
        case DINNER:
            expenseName = "Dinner";
            break;
        case BREAKFAST:
            expenseName = "Breakfast";
            break;
        case CAR_RENTAL:
            expenseName = "Car Rental";
            break;
        }

        string mealOverExpensesMarker = (expense->type == DINNER && expense->amount > 5000) || (expense->type == BREAKFAST && expense->amount > 1000) ? "X" : " ";

        if (htmlMode) {
            cout << "<tr><td>" << expenseName << "</td><td>" << expense->amount << "</td><td>" << mealOverExpensesMarker << "</td></tr>\n";
        } else {
            cout << expenseName << '\t' << expense->amount << '\t' << mealOverExpensesMarker << '\n';
        }

        total += expense->amount;
    }
    if (htmlMode) {
        cout << "</tbody>" << '\n';
        cout << "</table>" << '\n';
    }

    if (htmlMode) {
        cout << "<p>Meal expenses: " << mealExpenses << "</p>\n";
        cout << "</p>Total expenses: " << total << "</p>\n";
    } else {
        cout << "Meal expenses: " << mealExpenses << '\n';
        cout << "Total expenses: " << total << '\n';
    }

    if (htmlMode) {
        cout << "</body>" << '\n';
        cout << "</html>" << '\n';
    }
}
