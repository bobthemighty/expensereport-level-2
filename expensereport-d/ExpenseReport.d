import std.stdio;
import std.datetime.systime;

enum ExpenseType { DINNER, BREAKFAST, CAR_RENTAL }

struct Expense {
    ExpenseType type;
    int amount;
}

void printReport(Expense[] expenses, bool htmlMode) {
    int meals = 0;
    int total = 0;
    if (htmlMode) {
        writefln("<!DOCTYPE html>");
        writefln("<html lang=\"en\">");
        writefln("<head>");
        writefln("<title>Expense Report %s</title>", Clock.currTime());
        writefln("</head>");
        writefln("<body>");
        writefln("<h1>Expense Report %s</h1>", Clock.currTime());
    } else {
        writefln("Expenses: %s", Clock.currTime());
    }

    if (htmlMode) {
        writefln("<table>");
        writefln("<tbody>");
        writefln("<tr><th span=\"col\">Type</th><th span=\"col\">Amount</th><th span=\"col\">Over Limit</th></tr>");
    }
    foreach (expense; expenses) {
        if (expense.type == ExpenseType.DINNER || expense.type == ExpenseType.BREAKFAST) {
            meals += expense.amount;
        }
        string name = "";
        switch (expense.type) {
        case ExpenseType.DINNER: name = "Dinner"; break;
        case ExpenseType.BREAKFAST: name = "Breakfast"; break;
        case ExpenseType.CAR_RENTAL: name = "Car Rental"; break;
        default: name = "";
        }
        string mealOverExpensesMarker = expense.type == ExpenseType.DINNER && expense.amount > 5000 || expense.type == ExpenseType.BREAKFAST && expense.amount > 1000 ? "X" : " ";
        if (htmlMode) {
            writefln("<tr><td>%s</td><td>%s</td><td>%s</td></tr>", name, expense.amount, mealOverExpensesMarker);
        } else {
            writefln("%s\t%s\t%s", name, expense.amount, mealOverExpensesMarker);
        }
        total += expense.amount;
    }
    if (htmlMode) {
        writefln("</tbody>");
        writefln("</table>");
    }

    if (htmlMode) {
        writefln("<p>Meal expenses: %d</p>", meals);
        writefln("<p>Total expenses: %d</p>", total);
    } else {
        writefln("Meal expenses: %d", meals);
        writefln("Total expenses: %d", total);
    }

    if (htmlMode) {
        writefln("</body>");
        writefln("</html>");
    }
}

void main() {
    Expense[] expenses = [
        { ExpenseType.DINNER, 5000 },
        { ExpenseType.DINNER, 5001 },
        { ExpenseType.BREAKFAST, 1000 },
        { ExpenseType.BREAKFAST, 1001 },
        { ExpenseType.CAR_RENTAL, 4 },
    ];
    printReport(expenses, true);
}
