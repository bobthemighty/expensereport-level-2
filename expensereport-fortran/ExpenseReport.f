      MODULE ExpenseReport
        IMPLICIT NONE
        ENUM, BIND(C)
          ENUMERATOR :: BREAKFAST, DINNER, CAR_RENTAL
        ENDENUM

        TYPE :: expense
          INTEGER :: type
          INTEGER :: amount
        END TYPE

        CONTAINS
        SUBROUTINE printReport(htmlMode, expenses)
          INTEGER :: htmlMode
          TYPE(expense), DIMENSION (:), ALLOCATABLE :: expenses
          INTEGER :: total = 0
          INTEGER :: mealExpenses = 0
          INTEGER :: i
          CHARACTER(LEN = 20) :: expenseName
          CHARACTER(LEN = 1) :: mealOverExpensesMarker
          IF (htmlMode == 1) THEN
            PRINT "(a)", '<!DOCTYPE html>'
            PRINT "(a)", '<html lang="en">'
            PRINT "(a)", '<head>'
            PRINT "(a)", '<title>Expense Report</title>'
            PRINT "(a)", '</head>'
            PRINT "(a)", '<body>'
            PRINT "(a)", '<h1>Expense Report</h1>'
          ELSE
            PRINT "(a)", 'Expense Report'
          END IF

          IF (htmlMode == 1) THEN
            PRINT "(a)", '<table>'
            PRINT "(a)", '<thead>'
            PRINT "(a)", '<tr>'
            PRINT "(a)", '<th scope="col">Type</th>'
            PRINT "(a)", '<th scope="col">Amount</th>'
            PRINT "(a)", '<th scope="col">Over Limit</th>'
            PRINT "(a)", '</tr>'
            PRINT "(a)", '</thead>'
            PRINT "(a)", '<tbody>'
          END IF
          DO i = LBOUND(expenses, 1), UBOUND(expenses, 1)
            IF (expenses(i)%type == BREAKFAST .OR. expenses(i)%type == DINNER) &
              mealExpenses = mealExpenses + expenses(i)%amount
            SELECT CASE (expenses(i)%type)
              CASE (DINNER)
                expenseName = "Dinner"
              CASE (BREAKFAST)
                expenseName = "Breakfast"
              CASE (CAR_RENTAL)
                expenseName = "Car Rental"
            END SELECT
            IF (expenses(i)%type == BREAKFAST .AND. expenses(i)%amount > 1000 .OR. &
                expenses(i)%type == DINNER .AND. expenses(i)%amount > 5000) THEN
              mealOverExpensesMarker = "X"
            ELSE
              mealOverExpensesMarker = " "
            END IF
            total = total + expenses(i)%amount
            IF (htmlMode == 1) THEN
              PRINT "(a,a,a,1x,i10,a,1x,a,a)", &
                '<tr><td>', &
                expenseName, &
                '</td><td>', &
                expenses(i)%amount, &
                '</td><td>', &
                mealOverExpensesMarker, &
                '</td></tr>'
            ELSE
              PRINT "(a,1x,i10,1x,a)", expenseName, expenses(i)%amount, mealOverExpensesMarker
            END IF
          END DO
          IF (htmlMode == 1) THEN
            PRINT "(a)", '</tbody>'
            PRINT "(a)", '</table>'
          END IF

          IF (htmlMode == 1) THEN
            PRINT "(a,i10,a)", '<p>Meal expenses: ', mealExpenses, &
              '</p>'
            PRINT "(a,i10,a)", '<p>Total: ', total, '</p>'
          ELSE
            PRINT "(a,i10)", 'Meal expenses: ', mealExpenses
            PRINT "(a,i10)", 'Total: ', total
          END IF

          IF (htmlMode == 1) THEN
            PRINT "(a)", '</body>'
            PRINT "(a)", '</html>'
          END IF
        END SUBROUTINE printReport

      END MODULE ExpenseReport

      PROGRAM main
        USE ExpenseReport
        IMPLICIT NONE
        TYPE(expense), DIMENSION (:), ALLOCATABLE :: expenses
        ALLOCATE(expenses(5))
        expenses(1)%type = BREAKFAST
        expenses(1)%amount = 1000
        expenses(2)%type = BREAKFAST
        expenses(2)%amount = 1001
        expenses(3)%type = DINNER
        expenses(3)%amount = 5000
        expenses(4)%type = DINNER
        expenses(4)%amount = 5001
        expenses(5)%type = CAR_RENTAL
        expenses(5)%amount = 4
        CALL printReport(1, expenses)
        DEALLOCATE(expenses)
      END PROGRAM main
