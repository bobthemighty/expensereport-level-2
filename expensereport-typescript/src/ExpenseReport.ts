type ExpenseType = "dinner" | "breakfast" | "car-rental"

class Expense {
  constructor(public type: ExpenseType, public amount: number) {
  }

  get name() : string {
    switch(this.type) {
      case "dinner":
        return "Dinner"
      case "breakfast":
        return "Breakfast"
      case "car-rental":
        return "Car Rental"
    }
  }

  get isOverLimits() : boolean {
    return this.type == "dinner" && this.amount > 5000 || this.type == "breakfast" && this.amount > 1000 ? true: false
  }
}

class Expenses implements Iterable<Expense> {
  constructor (private items: Array<Expense>){}

  [Symbol.iterator]() : Iterator<Expense> {
    return this.items[Symbol.iterator]();
  }

  get mealExpenses() {
    return this.items.filter(x => x.type !== "car-rental").reduce((acc, x) => acc + x.amount, 0)
  }

  get totalExpenses() {
    return this.items.reduce((acc, x) => acc + x.amount, 0)
  }
}

class PlaintextWriter {
  writeHeader() {
    process.stdout.write("Expense Report: " + new Date().toISOString().substr(0, 10) + "\n")
  }

  writeItems(expenses: Iterable<Expense>) {
    for (const expense of expenses) {
      process.stdout.write(expense.name + "\t" + expense.amount + "\t" + (expense.isOverLimits? "X": " ") + "\n")
    }
  }

  writeTotal(expenses: Expenses) {
    process.stdout.write("Meal Expenses: " + expenses.mealExpenses + "\n")
    process.stdout.write("Total Expenses: " + expenses.totalExpenses + "\n")
  }
}

function printReport(htmlMode: boolean, _expenses: Expense[]) {
  const expenses = new Expenses(_expenses);
  const writer = new PlaintextWriter();

  if(!htmlMode){
    writer.writeHeader()
    writer.writeItems(expenses)
    writer.writeTotal(expenses)
    return
  }

    process.stdout.write("<!DOCTYPE html>\n");
    process.stdout.write("<html>\n");
    process.stdout.write("<head>\n");
    process.stdout.write("<title>Expense Report: " + new Date().toISOString().substr(0, 10) + "</title>\n")
    process.stdout.write("</head>\n");
    process.stdout.write("<body>\n");
    process.stdout.write("<h1>Expense Report: " + new Date().toISOString().substr(0, 10) + "</h1>\n")

    process.stdout.write("<table>\n");
    process.stdout.write("<thead>\n");
    process.stdout.write("<tr><th scope=\"col\">Type</th><th scope=\"col\">Amount</th><th scope=\"col\">Over Limit</th></tr>\n");
    process.stdout.write("</thead>\n");
    process.stdout.write("<tbody>\n");


    for (const expense of expenses) {
      const mealOverExpensesMarker = expense.isOverLimits ? "X" : " "
      process.stdout.write("<tr><td>" + expense.name + "</td><td>" + expense.amount + "</td><td>" + mealOverExpensesMarker + "</td></tr>\n")
    }
    process.stdout.write("</tbody>\n");
    process.stdout.write("</table>\n");

    process.stdout.write("<p>Meal Expenses: " + expenses.mealExpenses + "</p>\n")
    process.stdout.write("<p>Total Expenses: " + expenses.totalExpenses + "</p>\n")

    process.stdout.write("</body>\n");
    process.stdout.write("</html>\n");
}

export {printReport, Expense, ExpenseType}
