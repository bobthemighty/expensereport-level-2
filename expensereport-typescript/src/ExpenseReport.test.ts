import { printReport, Expense } from './ExpenseReport'

describe(`ExpenseReport`, () => {
    it(`should keep its original behavior`, () => {
        let interceptedOutput = ""
        jest.spyOn(process.stdout, "write").mockImplementation((output: string): boolean => {
            interceptedOutput += output
            return true;
        })
        printReport(true, [
          new Expense("breakfast", 1000),
          new Expense("breakfast", 1001),
          new Expense("dinner", 5000),
          new Expense("dinner", 5001),
          new Expense("car-rental", 4),
        ])
        expect(interceptedOutput).toMatchSnapshot()
    })
})
