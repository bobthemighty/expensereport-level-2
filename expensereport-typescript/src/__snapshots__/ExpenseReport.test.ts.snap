// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseReport should keep its original behavior 1`] = `
"<!DOCTYPE html>
<html>
<head>
<title>Expense Report: 2022-06-29</title>
</head>
<body>
<h1>Expense Report: 2022-06-29</h1>
<table>
<thead>
<tr><th scope=\\"col\\">Type</th><th scope=\\"col\\">Amount</th><th scope=\\"col\\">Over Limit</th></tr>
</thead>
<tbody>
<tr><td>Breakfast</td><td>1000</td><td> </td></tr>
<tr><td>Breakfast</td><td>1001</td><td>X</td></tr>
<tr><td>Dinner</td><td>5000</td><td> </td></tr>
<tr><td>Dinner</td><td>5001</td><td>X</td></tr>
<tr><td>Car Rental</td><td>4</td><td> </td></tr>
</tbody>
</table>
<p>Meal Expenses: 12002</p>
<p>Total Expenses: 12006</p>
</body>
</html>
"
`;
